### DISCUSSION TITLE
v4 sync vs async preference

### DISCUSSION THREAD

**Discussion by kennethjyang** [isar/isar#1413] [Help] [answered]
> All operations are now synchronous by default and there are Async methods for asynchronous operations.

Does this mean that it is recommended to use synchronous calls instead of async ones? And if so, why? I do like this change, I just find it a bit strange considering most other data storage solutions all use asynchronous calls.

Also, how can I avoid Future propagation when the first call to open an Isar database requires something like `await getApplicationDocumentsDirectory()`? If all other operations are synchronous, it would be great if the first step could also be synchronous so subsequent operations don't also have to be futures. My current solution is to wrap my whole app in a Future provider and I would like to know if there is a more elegant solution to this.

**Comment by giga10** [comment]
Same two things are bothering me too. @simc

**Comment by simc** [comment]
Most other Flutter databases including sqflite, realm and objectbox use synchronous operations.

With Isar you can choose. Getting the path is a one time operation and should not cause future propagation.

`getApplicationDocumentsDirectory() `is not an Isar method and I cannot make it synchronous. Feel free to use another way to fetch a valid path.

**Reply by kjy5** [reply]
Now that the defaults are synchronous, does this mean it's the recommended style?

**Reply by simc** [reply, marked as answer]
Both options have advantages and disadvantages. I recommend to try synchronous methods and only if you experience performance issues switch to async

**Reply by kennethjyang** [reply]
Got it. Thanks!

**Reply by Kargthefergus** [reply]
Simon - do you have any ideas when v4 might be released?  I have an app with two dbs that need to be open simultaneously, and have received the error for that.  I would rather not rewrite if there is going to be an easier way to upgrade the code.  Thanks for doing what you do!

**Comment by tomasweigenast** [comment]
> Both options have advantages and disadvantages. I recommend to try synchronous methods and only if you experience performance issues switch to async

Can you tell use what are the advantages and disadvantages for each option? Thanks
